### DISCUSSION TITLE
Route.resource enhancement idea

### DISCUSSION THREAD

**Discussion by bitkidd** [adonisjs/core#2655] [Ideas]
Hey!
Recently was working on a project of mine and got an idea that `Route.resource` can be upgraded significantly by allowing it to be nested and allowing to specify a custom `:param`.

What it may add us, It may allow us to write routes like this:
```js
Route.resource('posts', 'PostsController', () => {
  Route.resource('comments', 'CommentsController', { param: 'uid' }).apiOnly()
}, { param: 'uid' }).apiOnly()
```

This may generate these routes:
| Method | Route | Handler |
| --- | --- | --- |
| GET | /posts | PostsController.index |
| GET | /posts/:uid | PostsController.show |
| POST | /posts | PostsController.store |
| PUT, PATCH | /posts/:uid | PostsController.update |
| DELETE | /posts/:uid | PostsController.destroy |
| GET | /posts/:post_uid/comments | CommentsController.index |
| GET | /posts/:post_uid/comments/:uid | CommentsController.show |
| POST | /posts/:post_uid/comments | CommentsController.store |
| PUT, PATCH | /posts/:post_uid/comments/:uid | CommentsController.update |
| DELETE | /posts/:post_uid/comments/:uid | CommentsController.destroy |

Not sure about the design, as it makes code a bit junky, maybe it would be cooler to be able to do something like this:
```js
Route.resource('posts', 'PostsController', () => {
  Route.resource('comments', 'CommentsController').param('uid').apiOnly()
}).param('uid').apiOnly()
```

What do you think?

**Comment by RomainLanz** [comment]
Hey @bitkidd! 👋 

You can already have nested resources.

```ts
Route.resource('posts', 'PostsController').apiOnly()
Route.resource('posts.comments', 'CommentsController').apiOnly()
```

📚  https://docs.adonisjs.com/guides/controllers#nested-resources

--

What would be the use-case to change `params` name?

**Reply by bitkidd** [reply]
Yes, I saw the docs, but in my own experience for public routes I tend not to disclose database ids.
So then I just recreate all the routes by hand specifying `params`.

It is not a very dramatic thing really, just an idea, as Rails has it and allows to define routes kinda like this:
```ruby
resources :posts, param: :uid do
  resources :comments, param: :uid
end
```

Not sure about Laravel though, but it may make life a bit, a tiny bit better 😄

**Reply by RomainLanz** [reply]
Nothing stops you from using `params.id` (or anything else) to map to an `uuid` row (or anything else).
This mapping is done inside your controller when querying the model.
